packages/compass/src/main/auto-update-manager.ts

```diff
@@ -109,52 +135,143 @@ export const enum AutoUpdateManagerState {
   ReadyToUpdate = 'ready-to-update',
   Restarting = 'restarting',
   RestartDismissed = 'restart-dismissed',
+  Idle = 'idle',
 }
 
 const FOUR_HOURS = 1000 * 60 * 60 * 4;
 const THIRTY_SECONDS = 30_000;
 
 type StateUpdateAction = (
-  this: AbortSignal,
-  updateManager: typeof CompassAutoUpdateManager,
+  this: typeof CompassAutoUpdateManager,
   ...args: any[]
-) => void | Promise<void>;
+) => Generator;
 
-const manualCheck: StateUpdateAction = function (updateManager) {
-  updateManager.setState(AutoUpdateManagerState.CheckingForUpdates, true);
+const manualCheck: StateUpdateAction = function* () {
+  yield void this.setState(AutoUpdateManagerState.CheckingForUpdates, true);
```

Comment:
Maybe I’m missing something here, but this seems contradictory – we’re using `yield` like an improved `await` here, but if that’s the case, then it doesn’t make sense to `void` an expression that we pass to it?

Reply:
You're right, and generally I think this particular handler is confusing because of how I typed this whole thing: while we can yield a promise here, it doesn't really make a difference because this action just kicks off another state transition, but I do it because the way I typed `StateUpdateAction` it needs to be a generator and our eslint doesn't like generators without yields. 

I think I want to change `setState` to return `never` instead of a promise to highlight that it's just thing that starts a new state transition and "interrupts" the currently running one, I think `never` would communicate this well.

Does this sound good to you? 🤔

Reply:
Yeah, `setState()` never returns when called from inside one of the state update actions because it aborts and therefore the `yield` throws, right? That makes sense then (although I think you want `Promise<never>` and not `never` because you still need to `yield` in order to actually get that exception?)

Reply:
Yes, totally!

Reply:
Wait, actually I think I confused myself on Friday (which is tbh my main concern with this refactor 😆), I don't really need to make this function to be async, I only need to yield here, at the moment the `flow` function will be checking the signal, the in-progress operation will be (syncronously) aborted. Or did you mean strictly in the types that it returns here?

Also as a side-note: wow typing generators is kinda painful 😳

Reply:
> Or did you mean strictly in the types that it returns here?

Yeah, it’s … odd that the function returns but doesn’t _really_ when used from inside one of these functions here.

> typing generators is kinda painful :flushed:

Yes – I know it would require passing around more state here, but I am kind of wondering if it wouldn’t be easier to use regular async/await with a function to call that throws when there is an interrupt (we do this in mongosh, for example: https://github.com/mongodb-js/mongosh/blob/825d75e068ac0a24ac06c474a0d39cf29fb4f168/packages/snippet-manager/src/snippet-manager.ts#L173 – it’s more verbose but might also be a bit easier to follow when reading the code)

Reply:
Yeah, that's something I wanted to avoid really, but also don't want to make this unmaintainable. I'll start from scratch and will just similar checkers after every line in state updaters